feat(icons): add deprecationReason and alias deprecation feature

karsa-mistmere opened the pull request:
closes https://github.com/lucide-icons/lucide/issues/1957

## What is the purpose of this pull request?
- [x] New Feature

### Description
When deprecating icons (looking at you, brand icons) or aliases, it is important to clearly mark the reason of deprecation as well as any steps the users should take, this PR adds the possibility to add the reason for deprecation to icons as well as the possibility to mark aliases as deprecated as per #1957 

## Before Submitting <!-- For every PR! -->
<!-- All of these requirements must be fulfilled. -->
- [x] I've read the [Contribution Guidelines](https://github.com/lucide-icons/lucide/blob/main/CONTRIBUTING.md).
- [x] I've checked if there was an existing PR that solves the same issue.

jguddas commented on icon.schema.json:
```diff
@@ -21,7 +21,26 @@
+          {
+            "type": "string"
+          },
+          {
+            "type": "object",
+            "properties": {
+              "name": {
+                "type": "string"
+              },
+              "deprecated": {
+                "type": "boolean",
+                "default": false
+              },
+              "deprecationReason": {
+                "type": "string"
+              }
+            }
+          }
```
I would say we should enforce that you need to provide a deprecation reason.
And also enforce that you only can set a depracation reason when you set deprecated to true.
```suggestion
          {
            "type": "object",
            "required": [
              "name",
              "deprecate",
              "deprecationReason"
            ],
            "properties": {
              "name": {
                "type": "string"
              },
              "deprecated": {
                "const": true
              },
              "deprecationReason": {
                "type": "string"
              }
            }
          }
```

karsa-mistmere replied:
> I would say we should enforce that you need to provide a deprecation reason. And also enforce that you only can set a depracation reason when you set deprecated to true.

Good idea, although I think a pre commit hook if more feasible than enforcing it via JSON schema.

jguddas replied:
We can just run ajv on commit tho, we already do the lint run in the pr workflow.

karsa-mistmere replied:
Neato, I didn't realise we did, I've added a `dependentRequired` clause to both.
```json
  "dependentRequired": {
    "deprecated": ["deprecationReason"]
  }
```

jguddas commented on icon.schema.json:
```diff
@@ -51,6 +70,12 @@
     "deprecated": {
       "type": "boolean",
       "default": false
+    },
+    "deprecationReason": {
+      "type": "string"
+    },
```
We can also enforce the relation ship here.

karsa-mistmere replied:
I've added a `dependentRequired` clause on both instances.

jguddas commented on icons/between-horizontal-end.json:
```diff
@@ -33,6 +33,10 @@
+    {
+      "name": "between-horizonal-end",
+      "deprecated": true,
+      "deprecationReason": "Renamed because of typo. This export will be removed in v1.0"
```
I think adding this here feels wrong, I would prefer adding this in the generator.
I would vote to keep them after v1 in general, too much hassle for the user for not that much upside.
```suggestion
      "deprecationReason": "Renamed because of typo."
```

karsa-mistmere replied:
Ideally, after v1.0 deprecated aliases could be kept and automatically removed in the next major release following the one they were added in. If and when we get there I agree that the generator should add this information. For now I think this "dirty hack" should do, as the aforementioned functionality is just too much of a hassle.

jguddas commented:
My vote is to `deprecationReason` an enum instead of free text, so we can ensure that there are no typos and control what depracation types there are.

karsa-mistmere commented:
> My vote is to `deprecationReason` an enum instead of free text, so we can ensure that there are no typos and control what depracation types there are.

Any ideas as to what these options should be?

jguddas commented:
> > My vote is to `deprecationReason` an enum instead of free text, so we can ensure that there are no typos and control what depracation types there are.
> 
> Any ideas as to what these options should be?

```jsonc
// depracation-reasons.en.json
{
  "alias.typo": "Renamed because of typo, use {name} instead.",
  "icon.brand": "Brand icons have been deprecated and are due to be removed, we recommend using simpleicons.org instead, refer to https://github.com/lucide-icons/lucide/issues/670"
}
```
https://formatjs.io/docs/intl-messageformat

karsa-mistmere commented:
I think i18n is a bit of an overkill, since all of our packages are in English, but I've implemented a rudimentary template function and converted the deprecation reason to an enumeration.

(I meant options apart from these two, so that we are somewhat futureproof, but we can think about that later)

jguddas commented on icon.schema.json:
```diff
@@ -21,7 +21,32 @@
+              "deprecationReason": {
+                "$ref": "#/$defs/deprecationReasons"
+              },
+              "gracePeriod": {
```
I asked copilot for a better name since I had no clue from just looking at the variable name what it is gonna do.
```suggestion
              "toBeRemovedInVersion": {
```

jguddas replied:
Another option could be a boolean `toBeRemovedInNextMajorVersion` 🤷

karsa-mistmere replied:
Fine by me, renamed it.

jguddas commented on icon.schema.json:
```diff
@@ -49,8 +74,27 @@
+  "$defs": {
+    "deprecationReasons": {
+      "type": "string",
+      "enum": ["icon.brand", "alias.typo", "alias.name"]
```
```suggestion
      "enum": ["icon.guideline-violation.brand", "alias.typo", "alias.guideline-violation.number"]
```

jguddas replied:
And you should not be able to do `alias[0].depracationReason: "icon.bla"`, only "alias.bla" should be allowed, they should be seperated imo.

karsa-mistmere replied:
Done.

jguddas commented on tools/build-icons/utils/deprecationReasonTemplate.mjs:
```diff
@@ -0,0 +1,13 @@
+export function deprecationReasonTemplate(
+  deprecationReason,
+  { componentName, iconName, toBeRemovedInVersion },
+) {
```
```suggestion
) {
const removalNotice = toBeRemovedInVersion ? `This alias will be removed in ${toBeRemovedInVersion}` : undefined
```

ericfennis reviewed (approved):
Nice lovely work!

karsa-mistmere commented:
> What about my unanswered pr comments?

They've now been addressed.

jguddas commented on tools/build-icons/utils/deprecationReasonTemplate.mjs:
```diff
@@ -0,0 +1,17 @@
+    case 'alias.typo':
+      return `Renamed because of typo, use {@link ${componentName}} instead.${removalNotice}`;
+    case 'alias.naming':
+      return `This name is no longer compliant with our guidelines, use {@link ${componentName}} instead.${removalNotice}`;
```
```suggestion
      return `This icon has been renamed because it is no longer compliant with our guidelines, use {@link ${componentName}} instead.${removalNotice}`;
```

karsa-mistmere replied:
What about?

```suggestion
      return `The name of this icon had to be changed because it didn't meet our guidelines anymore, use {@link ${componentName}} instead.${removalNotice}`;
```

jguddas replied:
Maybe instead of `had to be changed` do `was changed`.
